mbedtls/src/rng/rdrand.rs

```diff
@@ -7,35 +7,71 @@
  * according to those terms. */
 
 use core::slice::from_raw_parts_mut;
+
 use mbedtls_sys::types::raw_types::{c_int, c_uchar, c_void};
 use mbedtls_sys::types::size_t;
 
+#[cfg(target_arch = "x86")]
+use core::arch::x86_64::{_rdrand32_step as _rdrand_step, _rdseed32_step as _rdseed_step};
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::{_rdrand64_step as _rdrand_step, _rdseed64_step as _rdseed_step};
+
+// Intel documentation claims that if hardware is working RDRAND will produce output
+// after at most 10 attempts
+const RDRAND_READ_ATTEMPTS: usize = 10;
+
+// Intel does not document the number of times RDSEED might consecutively fail, but in
+// example code uses 75 as the upper bound.
+const RDSEED_READ_ATTEMPTS: usize = 75;
+
+fn call_cpu_rng<T, F>(attempts: usize, intrin: unsafe fn(&mut T) -> i32, cast: F) -> Option<usize>
+where
+    T: Sized + Default,
+    F: FnOnce(T) -> usize,
+{
+    assert_eq!(core::mem::size_of::<T>(), core::mem::size_of::<usize>());
+
+    for _ in 0..attempts {
+        let mut out = T::default();
+        let status = unsafe { intrin(&mut out) };
+        if status == 1 {
+            return Some(cast(out));
+        }
+    }
+    None
+}
+
+fn rdrand() -> Option<usize> {
+    call_cpu_rng(RDRAND_READ_ATTEMPTS, _rdrand_step, |x| x as usize)
+}
+
+fn rdseed() -> Option<usize> {
+    call_cpu_rng(RDSEED_READ_ATTEMPTS, _rdseed_step, |x| x as usize)
+}
+
+fn write_rng_to_slice(outbuf: &mut [u8], rng: fn() -> Option<usize>) -> c_int {
```

Comment:
This is great, I was unhappy about the amount of code in the `unsafe` block that didn't need to be there.